### PR TITLE
feat(wallet): add timeout client connection

### PR DESCRIPTION
**Pull request by Ja7ad**
## Description

This PR introduces a client timeout for connections when dialing to the server. The primary goal is to prevent delays during query retrieval by setting a maximum duration for establishing a connection. This ensures that if a server is unresponsive or slow, the client will not hang indefinitely, thereby improving the overall responsiveness and reliability of the application.

### CODE REVIEW

**Review by kehiy — COMMENTED** [review]
Is this work in GUI wallet as well?

**Comment by kehiy** [comment]
@Ja7ad I've tested it manually with 1 second as timeout but it took 24 seconds to respond to me.

My command:
```sh
./pactus-wallet address balance pc1zvchrej8nwfgvsduw47r4qmss5u25kvls2fx8e6 --timeout 1 
```

File: wallet/client.go [hunk 1]
```diff
@@ -39,16 +43,30 @@ func (c *grpcClient) SetServerAddrs(servers []string) {
...
+		if len(c.servers) > 1 {
+			opts = append(opts, grpc.WithContextDialer(func(_ context.Context, s string) (net.Conn, error) {
+				return net.DialTimeout("tcp", s, c.timeout)
+			}))
+		}
+
+		conn, err := grpc.NewClient(server, opts...)
 		if err != nil {
-			continue
+			return err
```
Comment (kehiy) [inline wallet/client.go:69]: Here we must continue, we have a list of servers...
We can't stop the whole process if one of the servers is offline.

**Review by kehiy — CHANGES_REQUESTED** [review]

**Comment by Ja7ad** [comment]
> @Ja7ad I've tested it manually with 1 second as timeout but it took 24 seconds to respond to me.
> 
> My command:
> 
> ```shell
> ./pactus-wallet address balance pc1zvchrej8nwfgvsduw47r4qmss5u25kvls2fx8e6 --timeout 1 
> ```
> 
> Shouldn't it return a timeout error?

This timeout duration for dialing client, gRPC client is immediately and connecting the server happens in background.

When call `GetBlockchainInfo` client started to dialing and call RPC method.

**Comment by Ja7ad** [comment]
> @Ja7ad so, with only 7 server, and 1 second timeout after 7 second I have to get data or timeout or can't connect to servers.
> 
> Here is the servers list:
> 
> ```json
> "mainnet": [
>         "localhost:50051",
>         "bootstrap1.pactus.org:50051",
>         "bootstrap2.pactus.org:50051",
>         "bootstrap3.pactus.org:50051",
>         "bootstrap4.pactus.org:50051",
>         "65.109.234.125:50051",
>         "157.90.111.140:50051"
> ]
> ```
> 
> But it took 24 seconds in my case, am I wrong?

This timeout is when dialing not query.

**Comment by kehiy** [comment]
@Ja7ad our goal is to reduce the latency or just get a timeout. what do you think if we apply the timeout to queries as well? as we defined it on Trello task?

The rest looks good to me, we can merge it.

**Comment by Ja7ad** [comment]
> @Ja7ad our goal is to reduce the latency or just get a timeout. what do you think if we apply the timeout to queries as well? as we defined it on Trello task?
> 
> The rest looks good to me, we can merge it.

Our delay is on when dialing in the best server.

**Comment by kehiy** [comment]
@Ja7ad but based on my manual test, the process is still slow but there is no timeout error. which is not the same as our goal.

File: cmd/wallet/main.go [hunk 2]
```diff
@@ -44,6 +49,11 @@ func main() {
 		cmd.PactusDefaultWalletPath(cmd.PactusDefaultHomeDir()), "the path to the wallet file")
 	offlineOpt = rootCmd.PersistentFlags().Bool("offline", false, "offline mode")
 	serverAddrOpt = rootCmd.PersistentFlags().String("server", "", "server gRPC address")
+	timeoutOpt = rootCmd.PersistentFlags().Int("timeout", 1,
+		"specifies the timeout duration for the client connection in seconds. "+
+			"this timeout determines how long the client will attempt to establish a connection with the server "+
```
Comment (b00f) [inline cmd/wallet/main.go:54]: It is too long. How the `-h` looks like after this change?
Do we have "Long Help" option in cobra?

File: wallet/client.go [hunk 3]
```diff
@@ -20,15 +22,17 @@ type grpcClient struct {
...
 	transactionClient pactus.TransactionClient
 }
 
 func newGrpcClient() *grpcClient {
-	ctx := context.WithoutCancel(context.Background())
+	ctx := context.Background()
 
 	return &grpcClient{
 		ctx:               ctx,
+		timeout:           time.Second * 1,
```
Comment (b00f) [inline wallet/client.go:35]: Can't we get timeout when we instantiate the client?
```go 
func newGrpcClient(timeout int) {...}
````

File: wallet/client.go [hunk 4]
```diff
@@ -58,6 +76,10 @@ func (c *grpcClient) connect() error {
 		_, err = blockchainClient.GetBlockchainInfo(c.ctx,
 			&pactus.GetBlockchainInfoRequest{})
 		if err != nil {
+			if err := conn.Close(); err != nil {
```
Comment (b00f) [inline wallet/client.go:79]: I suggest to ignore error here:
```
_ = conn.Close()
```

File: wallet/wallet.go [hunk 5]
```diff
@@ -156,6 +156,10 @@ func (w *Wallet) SetServerAddr(addr string) {
 	w.grpcClient.SetServerAddrs([]string{addr})
 }
 
+func (w *Wallet) SetClientTimeout(timeout time.Duration) {
```
Comment (b00f) [inline wallet/wallet.go:159]: Set timoeout when creating wallet. Then we don't need this setter.

File: wallet/client.go [hunk 6]
```diff
@@ -39,14 +43,28 @@ func (c *grpcClient) SetServerAddrs(servers []string) {
...
 		return nil
 	}
 
 	for _, server := range c.servers {
-		conn, err := grpc.NewClient(server,
-			grpc.WithTransportCredentials(insecure.NewCredentials()))
+		opts := make([]grpc.DialOption, 0)
+		opts = append(opts, grpc.WithTransportCredentials(insecure.NewCredentials()))
+
+		if len(c.servers) > 1 {
```
Comment (b00f) [inline wallet/client.go:61]: There is one concern here:
1- User specify a server and timeout at the same time. This condition will not apply the timeout.
2- If `WalletManager`  creates wallet, it try to connect to `localhost`. In this case we may give large number for timeout.

**Review by b00f — CHANGES_REQUESTED** [review]

File: cmd/wallet/main.go (hunk 2 above)
Comment (b00f) [reply cmd/wallet/main.go:54]: I checked the Cobra. The 3rd argument in Options is `Usage`. So if we need to add Long Help we should add it into the command. It supports Long Help .

Comment (Ja7ad) [reply cmd/wallet/main.go:54]: > I checked the Cobra. The 3rd argument in Options is `Usage`. So if we need to add Long Help we should add it into the command. It supports Long Help .

This is switch not flag, just have usage.

**Comment by Ja7ad** [comment]
@b00f done

**Review by b00f — APPROVED** [review]

**Pull request merged**